### PR TITLE
[1/N] Initial prototype for multi-modal processor

### PR DESCRIPTION
**Pull request by DarkLight1337**
This PR adds the core code for multi-modal processor while maintaining backward compatibility. The main purpose of this PR is to "reserve" code changes (mainly related to the dependencies of multi-modal processor) to reduce the risk of merge conflicts caused by subsequent PRs.

Note that currently there are no models that use the new multi-modal processor - I will implement a few of them in the next PR. As such, the details of multi-modal processor are still subject to change.

Part of #10114

### CODE REVIEW

File: vllm/multimodal/processing.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    assert end_id == end_id_
+
+    # Remove duplicates (need to convert to tuple to be hashable)
+    unique_candidates = {
+        tuple(c)
+        for c in [candidate_basic, candidate_a, candidate_ab, candidate_b]
+    }
+
+    # Convert back to list
+    return [list(c) for c in unique_candidates]
```
Comment (DarkLight1337) [inline vllm/multimodal/processing.py:168]: Based on our existing code for Phi-3V

File: vllm/multimodal/processing.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    placeholder_length = len(placeholder_ids)
+
+    for start_idx in range(len(token_ids) - placeholder_length + 1):
+        if token_ids[start_idx:placeholder_length] == placeholder_ids:
+            token_ids[start_idx:placeholder_length] = get_replacement_ids()
+
+            return PlaceholderRange(offset=start_idx,
+                                    length=placeholder_length)
+
+    return None
```
Comment (DarkLight1337) [inline vllm/multimodal/processing.py:191]: Based on our existing code for HF Pixtral

File: vllm/multimodal/image.py [hunk 3]
```diff
@@ -41,15 +41,11 @@ def _get_hf_image_processor(
...
-        data: MultiModalData[object],
+        data: MultiModalData[ImageItem],
         **mm_processor_kwargs,
     ) -> MultiModalKwargs:
         model_config = ctx.model_config
 
-        # Processed by input processor
-        if isinstance(data, BatchFeature):
-            return MultiModalKwargs(data.data)
-
```
Comment (DarkLight1337) [inline vllm/multimodal/image.py:52]: This was added by the Llama-3.2 PR, but I found that this model doesn't use HF processor in vLLM input processor anymore, so it should be safe to remove.

**Comment by youkaichao** [comment]
@WoosukKwon  is implementing multi-modality inference in V1, and can provide better review here. I am not familiar with this code path.

File: vllm/inputs/preprocess.py [hunk 4]
```diff
@@ -243,6 +295,14 @@ def _prompt_to_llm_inputs(
             multi_modal_data = tokens_content.get("multi_modal_data")
             mm_processor_kwargs = tokens_content.get("mm_processor_kwargs")
 
+            if multi_modal_data is not None and self._can_process_multimodal():
+                return self._process_multimodal(
+                    prompt_token_ids,
+                    multi_modal_data,
+                    mm_processor_kwargs,
+                    lora_request=lora_request,
+                )
```
Comment (Isotr0py) [inline vllm/inputs/preprocess.py:304]: Should we add a fallback to remind user multimodal process is skipped if they inputs `multi_modal_data` but model doesn't haved a registered processor？

Comment (DarkLight1337) [reply vllm/inputs/preprocess.py:304]: Sure!

Comment (DarkLight1337) [reply vllm/inputs/preprocess.py:304]: Since none of the models use processor yet, I'll just log an info message.

**Review by Isotr0py — APPROVED** [review]
LGTM!

**Comment by Isotr0py** [comment]
Seems that v1 entrypoint test is failing: https://buildkite.com/vllm/fastcheck/builds/7789#01931e26-10d6-4960-b792-e949972b2aef/188-1114

**Comment by DarkLight1337** [comment]
@mergify.io rebase

**Comment by DarkLight1337** [comment]
@mergify rebase

**Pull request merged**